Exception naming sniffs

SenseException opened the pull request:
I've started to create sniffs for exception naming as described in the readme: `Abstract exception class names and exception interface names should be suffixed with Exception`.

This PR isn't a finished one yet and contains the first draft for naming exception interfaces. As this is about static code analysis, I had to make some assumptions about the parsed code.

## Interfaces

* If an interface has the name suffix `*Exception`, I assume, that the interface is an exception interface and needs to be handled by the sniff.
* If an interface extends an exception interface, I assume, that the interface has to be an exception and needs to be handled by the sniff.
* If an interface class extends `\Throwable`, it is definitely an exception and needs the `*Exception` suffix.
* If one of the extended interfaces is an exception, I assume that the interface extending them is an exception interface too.

This one is covered for most cases with the creation of the PR. There are still forms of interfaces I need to cover and the sniff isn't finished either, but there's still room for feedback on the existing code.

The sniff is currently not covering `use` cases, where class alias, comma separation and group use declaration is used, but the dependency slevomat/coding-standard has helper classes like `UseStatementHelper`, that can help parsing those. Should I use the classes of this dependency for doctrine's custom coding standard sniffs or should I not introduce code from dependencies there?

## Classes

~~This still needs to be handled with this PR.~~

* If an abstract class has the name suffix `*Exception`, I assume, that the class is an exception class and needs to be handled by the sniff.
* If a class extends an exception class or implements an exception interface, I assume, that the class has to be an exception and needs to be handled by the sniff.
* If one of the extended classes or implemented interfaces has the suffix `*Exception` in its name, I assume the class is an exception class and needs its name checked.

### Question

Should an abstract class only extend other abstract exception classes and basic PHP exception classes? Otherwise how do I know that an extended class without `*Exception` in its name is a valid regular exception class as the coding standard allows? Should I check if the class is in an exception namespace? Does the standard expect all exceptions in an exception namespace?

Addendum: leave exception classes have to be final. Exception namespaces are not in the scope of this PR.

SenseException committed:
Create sniff class base for exception interface naming

SenseException committed:
Install phpunit

SenseException committed:
Create tests for exception interface naming sniff

SenseException committed:
Add phpunit to the travis build

SenseException committed:
Make FooException of input-classes dirty

SenseException committed:
Adapt expected report to new phpcs result

greg0ire commented on phpunit.xml.dist:
```diff
@@ -0,0 +1,32 @@
+        3) To run the tests against the database type the following from within the
+           tests/ folder: phpunit -c <filename> ...
+           Example: phpunit -c mysqlconf.xml AllTests
+-->
```
Probably needs to be removed since this repo does not have to do with dbms'es

SenseException replied:
True. I guess we don't need the hint to copy the dist-file into a xml either, so I removed it.

SenseException committed:
Remove instruction comment

alcaeus commented:
> The sniff is currently not covering `use` cases, where class alias, comma separation and group use declaration is used, but the dependency slevomat/coding-standard has helper classes like `UseStatementHelper`, that can help parsing those. Should I use the classes of this dependency for doctrine's custom coding standard sniffs or should I not introduce code from dependencies there?

That depends on whether those classes are part of the public API: if they are, go for it. @kukulich could you clarify whether those helpers are public API covered under Semver rules?

> Should an abstract class only extend other abstract exception classes and basic PHP exception classes?

The assumption would make the sniff considerably easier, but strictly speaking, this isn't the case. I believe that we should take the shortcut and then expand the sniff once it's necessary to do so.

kukulich commented:
> could you clarify whether those helpers are public API covered under Semver rules?

Yes, they are :)

alcaeus commented:
Wonderful, thanks!

SenseException committed:
Use UseStatementHelper to parse imported fqcn

SenseException committed:
Handle alias for imported interfaces

SenseException committed:
Add test for comma separated namespaces

SenseException committed:
Add test for group use namespaces

SenseException committed:
Provide exception interface for phpcbf comparison test

Majkl578 commented on tests/Doctrine/Tests/Sniffs/Classes/ExceptionInterfaceNamingSniffTest.php:
```diff
@@ -0,0 +1,100 @@
+use PHP_CodeSniffer\Ruleset;
+use PHPUnit\Framework\TestCase;
+
+class ExceptionInterfaceNamingSniffTest extends TestCase
```
Please use `SlevomatCodingStandard\Sniffs\TestCase` as a base class and then its provided testing helpers, just like tests in Slevomat CS itself: https://github.com/slevomat/coding-standard/blob/master/tests/Sniffs/Types/EmptyLinesAroundTypeBracesSniffTest.php

Majkl578 commented on phpunit.xml.dist:
```diff
@@ -0,0 +1,20 @@
+        beStrictAboutChangesToGlobalState="true"
+        beStrictAboutOutputDuringTests="true"
+        beStrictAboutTestsThatDoNotTestAnything="true"
+        bootstrap="./vendor/squizlabs/php_codesniffer/tests/bootstrap.php"
```
Would be worth having custom `bootstrap.php` like https://github.com/slevomat/coding-standard/blob/master/tests/bootstrap.php.

Majkl578 commented on phpunit.xml.dist:
```diff
@@ -0,0 +1,20 @@
+        bootstrap="./vendor/squizlabs/php_codesniffer/tests/bootstrap.php"
+>
+
+    <testsuites>
```
This node isn't needed at all imho.

Majkl578 commented on phpunit.xml.dist:
```diff
@@ -0,0 +1,20 @@
+<?xml version="1.0" encoding="utf-8"?>
+<phpunit
```
Can you use phpunit configuration matching DBAL? https://github.com/doctrine/dbal/blob/master/phpunit.xml.dist#L14
Also no verbose by default please.

Majkl578 commented on lib/Doctrine/Sniffs/Classes/ExceptionInterfaceNamingSniff.php:
```diff
@@ -0,0 +1,110 @@
+    /**
+     * {@inheritdoc}
+     */
+    public function register()
```
`: array` / ` : iterable`

Majkl578 commented on lib/Doctrine/Sniffs/Classes/ExceptionInterfaceNamingSniff.php:
```diff
@@ -0,0 +1,110 @@
+    /**
+     * {@inheritdoc}
+     */
+    public function process(File $phpcsFile, $stackPtr)
```
`: void`

Majkl578 commented on lib/Doctrine/Sniffs/Classes/ExceptionInterfaceNamingSniff.php:
```diff
@@ -0,0 +1,110 @@
+            return;
+        }
+
+        if (! $hasExceptionName && ($isExtendingException || $isExtendingThrowable)) {
```
Please invert the logic and use _early return_ approach.

Majkl578 commented on lib/Doctrine/Sniffs/Classes/ExceptionInterfaceNamingSniff.php:
```diff
@@ -0,0 +1,110 @@
+        }
+
+        $extendedInterfaces = explode(',', $phpcsFile->getTokensAsString($start + 1, $limit - $start));
+        $extendedInterfaces = array_map('\trim', $extendedInterfaces);
```
no leading `\` here

SenseException replied:
I know that the coding-standard wants a dev to import functions with `use` and `trim()` is only used as a callback string here. Wouldn't removing the leading `\` lead to a lookup for `trim()` in the same namespace first?

Majkl578 replied:
No, string names are always fully qualified.

https://3v4l.org/HXbil

Majkl578 commented on lib/Doctrine/Sniffs/Classes/ExceptionInterfaceNamingSniff.php:
```diff
@@ -0,0 +1,110 @@
+
+        $isExtendingThrowable = (in_array(Throwable::class, $importedClassNames, true) &&
+            in_array(Throwable::class, $extendedInterfaces, true)) ||
+            in_array('\\Throwable', $extendedInterfaces, true);
```
No leading `\` should ever be used in const strings (the types are always to be FCQN).
If the `$extendedInterfaces` may contain names with `\`, it needs to be reworked (i.e. `use` resolution moved to `parseExtendedInterfaces()` directly).

SenseException replied:
Should I ignore the leading `\` then because another sniff should handle this smell?

Majkl578 replied:
The sniff should be able to determine whether `Throwable` is global type or namespaced type (either `use` or `\`), but it shouldn't be hardcoded here like this.

Majkl578 commented on lib/Doctrine/Sniffs/Classes/ExceptionInterfaceNamingSniff.php:
```diff
@@ -0,0 +1,110 @@
+
+        // Set original classname instead of alias
+        $extendedInterfaces = array_map(function (string $extendedInterface) use ($importedClassNames) {
+            if (isset($importedClassNames[$extendedInterface])) {
```
Whole block to be replaced by:
```php
return $importedClassNames[$extendedInterface] ?? $extendedInterface;
```

SenseException replied:
Good catch.

Majkl578 commented on lib/Doctrine/Sniffs/Classes/ExceptionInterfaceNamingSniff.php:
```diff
@@ -0,0 +1,110 @@
+        $extendedInterfaces = $this->parseExtendedInterfaces($phpcsFile, $stackPtr);
+
+        // Set original classname instead of alias
+        $extendedInterfaces = array_map(function (string $extendedInterface) use ($importedClassNames) {
```
needs return type

Majkl578 commented on .travis.yml:
```diff
@@ -20,6 +20,7 @@ install: travis_retry composer update --prefer-dist
 script:
   - vendor/bin/phpcs
   - vendor/bin/phpcs $(find tests/input/* | sort) --report=summary --report-file=phpcs.log; diff tests/expected_report.txt phpcs.log
+  - vendor/bin/phpunit
```
We will also need PHPStan since we now have real sniffs in this repo. :)

Majkl578 commented on lib/Doctrine/Sniffs/Classes/ExceptionInterfaceNamingSniff.php:
```diff
@@ -0,0 +1,110 @@
+use function preg_grep;
+use function strpos;
+
+class ExceptionInterfaceNamingSniff implements Sniff
```
`final`

Majkl578 reviewed (changes requested):
This sniff will be really helpful (not only) for ORM 3.0, DBAL 3.0, Migrations 2.0 and other libraries.

I'd also love to see a sniff that validates actual `throw` that we don't throw i.e. SPL exceptions. But that'd be separate task/sniff. :)

SenseException committed:
Refactor to a cleaner coding style

SenseException committed:
Improve parsing of interface names

SenseException committed:
Optimize phpunit config and create a custom bootstrap

SenseException committed:
Install phpstan for usage in travis build

SenseException committed:
Refactor tests to use SlevomatCodingStandard assertions

SenseException committed:
Move tests to different namespace for SlevomatCodingStandard sniff loading logic

SenseException committed:
Use early return on correct exception naming

SenseException commented:
@Majkl578 I've updated the PR for the interface handling.

SenseException committed:
Move test exception interfaces into subdirectory

SenseException committed:
Add type cast for possible nullable return value

SenseException committed:
Create tests for exception class sniff

SenseException committed:
Create first exception class sniff draft

SenseException committed:
Add tests for Throwable in same namespace

SenseException committed:
Move parsing of use statements into helper class

SenseException commented on lib/Doctrine/Helpers/UseStatementHelper.php:
```diff
@@ -0,0 +1,24 @@
+use PHP_CodeSniffer\Files\File;
+use SlevomatCodingStandard\Helpers\UseStatementHelper as UseHelper;
+
+class UseStatementHelper
```
I'm not a fan of `Helper` classes or a helper-namespace, but I kept naming and usage of static methods for consistency. If we should use a different approach for outsourcing logic, please let me know.

SenseException replied:
I'm not a fan of `Helper` classes or a helper-namespace, but I kept naming and usage of static methods for consistency. If we should use a different approach for outsourcing logic, please let me know.

Majkl578 commented:
I'll try to provide a more detailed review in the afternoon, here are just a few issues I spotted:
* namespace should be `Doctrine\CodingStandard`, or if not possible, only `DoctrineCodingStandard`
* please make sure `tests/` is covered by our CS too (except the tested files themselves) - this will point out some CS issues already there (missing declare()s, redundant `@param`. missing return types, missing spaces around concatenation)
* try to upgrade to PHPStan 0.10 with phpstan-strict-rules, putting configuration into phpstan.neon.dist

If you want, I can provide some _base_ for writing/testing sniffs, similar to how slevomat/coding-standard and cdn77/coding-standard are currently tested.
We should also finally rework the way we test our CS (not in this PR), because the current way of maintaining `tests/input` + `tests/expected_report.txt` is really messy and not maintainable. :)

Feel free to ping me on Slack if you want to discuss any of above or anything else related to the CS. 👍

SenseException committed:
Set codesniffer coding style for tests

SenseException committed:
Set phpstan 0.10 as minimum requirement

SenseException committed:
Create phpstan configuration

SenseException commented on tests/Doctrine/Sniffs/Classes/ExceptionClassNamingSniffTest.php:
```diff
@@ -1,59 +1,63 @@
     }
 
+    /**
+     * @return string[][]|int[][]
```
How does the correct syntax for multidimensional arrays look like? Seems like the sniff is satisfied with multiple possible formats.

SenseException committed:
Rename namespace root to DoctrineCodingStandard

SenseException committed:
Move logic of check for Throwable to UseStatementHelper

SenseException committed:
Move test classes into one file for valid interface naming sniff

SenseException commented:
Addition to Exception classes: They have to be `final`.

SenseException committed:
Move test classes into one file for valid class naming sniff

SenseException committed:
Check for non-abstract exceptions to be final

SenseException committed:
Move test classes into one file for invalid class naming sniff

SenseException committed:
Move test classes into one file for invalid interface naming sniff

SenseException committed:
Extend fixture classes for test cases

Ocramius commented:
> Addition to Exception classes: They have to be `final`.

Disagree on exception classes being `final` - unless we really segregate all exception *types* to interfaces (which would be OK), exceptions do in fact match the typical inheritance-first approach.

This means that each exception is required to implement at least 2 interfaces.

Majkl578 commented:
> Disagree on exception classes being final - unless we really segregate all exception types to interfaces (which would be OK), exceptions do in fact match the typical inheritance-first approach.

I'd say the ideal solution would be something like this:
* _leaf_ exceptions must be `final`
* _leaf_ exceptions must either:
  * extend an abstract ancestor exception, or
  * extend a root (SPL) exception and implement an exception marker interface
* exception marker interfaces must either:
  * extend Throwable, or
  * extend another marker interface from the same or any upper namespace
* _intermediate_ (abstract) exceptions should be avoided, if used, they must either:
  * extend a root (SPL) exception and implement an exception marker interface, or
  * extend another intermediate (abstract) exception

Interface-based hiearchy should be generally always preferred and usage of intermediate (abstract) exceptions should be rather exceptional.
Example:
```
Throwable
┗ [IM] Doctrine\DBAL\Exception\DbalException
       ┗ [IM] Doctrine\DBAL\Driver\Exception\DriverException

RuntimeException + Doctrine\DBAL\Driver\Exception\DriverException
┗ [A] Doctrine\DBAL\Driver\Exception\ContextualException (w/ i.e. `getSQLState()`)
        ┗ [L] Doctrine\DBAL\Driver\Exception\FailedQuery

[I] = exception interface
[M] = marker interface
[A] = intermediate (abstract) exception
[L] = leaf (final) exception
```

Majkl578 commented on lib/DoctrineCodingStandard/Sniffs/Classes/ExceptionInterfaceNamingSniff.php:
```diff
@@ -0,0 +1,97 @@
+    private function parseExtendedInterfaces(File $phpcsFile, int $stackPtr) : array
+    {
+        $limit = $phpcsFile->findNext([T_OPEN_CURLY_BRACKET], $stackPtr) - 1;
+        $start = $phpcsFile->findNext([T_EXTENDS], $stackPtr, $limit);
```
Use `SlevomatCodingStandard\Helpers\TokenHelper::findNext($phpcsFile, [T_EXTENDS], $stackPtr + 1, $limit)`.

Majkl578 commented on lib/DoctrineCodingStandard/Sniffs/Classes/ExceptionInterfaceNamingSniff.php:
```diff
@@ -0,0 +1,97 @@
+     */
+    private function parseExtendedInterfaces(File $phpcsFile, int $stackPtr) : array
+    {
+        $limit = $phpcsFile->findNext([T_OPEN_CURLY_BRACKET], $stackPtr) - 1;
```
Use `SlevomatCodingStandard\Helpers\TokenHelper::findNext($phpcsFile, [T_OPEN_CURLY_BRACKET], $stackPtr + 1)`.

Majkl578 commented on lib/DoctrineCodingStandard/Sniffs/Classes/ExceptionInterfaceNamingSniff.php:
```diff
@@ -0,0 +1,97 @@
+        $limit = $phpcsFile->findNext([T_OPEN_CURLY_BRACKET], $stackPtr) - 1;
+        $start = $phpcsFile->findNext([T_EXTENDS], $stackPtr, $limit);
+
+        if ($start === false) {
```
With the changes above: `=== null`.

Majkl578 commented on lib/DoctrineCodingStandard/Sniffs/Classes/ExceptionInterfaceNamingSniff.php:
```diff
@@ -0,0 +1,97 @@
+            return [];
+        }
+
+        $extendedInterfaces = explode(',', $phpcsFile->getTokensAsString($start + 1, $limit - $start));
```
It'd be better to do this properly, using combination of `SlevomatCodingStandard\Helpers\TokenHelper::findNextLocal($phpcsFile, [T_COMMA], $start, $limit)` + `SlevomatCodingStandard\Helpers\TokenHelper::getContent($phpcsFile, $start, $localEnd)` between `$start`/comma and comma/`$limit` (can be done with `for` or `do...whie`.

Majkl578 commented on lib/DoctrineCodingStandard/Sniffs/Classes/ExceptionInterfaceNamingSniff.php:
```diff
@@ -0,0 +1,97 @@
+    /**
+     * @return string[]
+     */
+    private function parseExtendedInterfaces(File $phpcsFile, int $stackPtr) : array
```
Can we move this to Helpers?

Majkl578 commented on lib/DoctrineCodingStandard/Sniffs/Classes/ExceptionInterfaceNamingSniff.php:
```diff
@@ -0,0 +1,97 @@
+    /**
+     * {@inheritdoc}
+     */
+    public function process(File $phpcsFile, $stackPtr) : void
```
Let's use sane names and name the pointer `$pointer`. :)

Majkl578 commented on lib/DoctrineCodingStandard/Sniffs/Classes/ExceptionInterfaceNamingSniff.php:
```diff
@@ -0,0 +1,97 @@
+    private const CODE_NOT_AN_EXCEPTION = 'NotAnException';
+
+    /**
+     * {@inheritdoc}
```
Drop, or rather document properly as `@return mixed[]`.

SenseException replied:
This would be `@return int[]` for this method. All `T_*` constants are integers.

Majkl578 replied:
Not all, some are strings too. (Can't remember which exactly.)

SenseException replied:
I didn't found ones that are strings. https://3v4l.org/aCWko
I'll keep `int[]`.

kukulich replied:
Constants defined by PHPCS itself are strings: https://github.com/squizlabs/PHP_CodeSniffer/blob/master/src/Util/Tokens.php#L12

Token constants in PHP are only integers.

Majkl578 replied:
Yep, those. :)

SenseException replied:
I added `string[]` for the PHPCS ones to the return type.

Majkl578 commented on lib/DoctrineCodingStandard/Sniffs/Classes/ExceptionInterfaceNamingSniff.php:
```diff
@@ -0,0 +1,97 @@
+    }
+
+    /**
+     * {@inheritdoc}
```
Document `@param int $pointer` instead.

SenseException replied:
Code sniffer doesn't like having that param annotation and wants me to use the int type hint in the method.

Majkl578 commented on lib/DoctrineCodingStandard/Sniffs/Classes/ExceptionInterfaceNamingSniff.php:
```diff
@@ -0,0 +1,97 @@
+            return $importedClassNames[$extendedInterface] ?? $extendedInterface;
+        }, $extendedInterfaces);
+
+        $hasExceptionName = preg_match('/Exception$/', (string) $phpcsFile->getDeclarationName($stackPtr)) === 1;
```
`$endsWithExceptionSuffix`

Majkl578 commented on lib/DoctrineCodingStandard/Sniffs/Classes/ExceptionInterfaceNamingSniff.php:
```diff
@@ -0,0 +1,97 @@
+
+        $isValidInterface = $hasExceptionName && ($isExtendingException || $isExtendingThrowable);
+        $isNoException    = ! $hasExceptionName && ! $isExtendingException && ! $isExtendingThrowable;
+        if ($isValidInterface || $isNoException) {
```
Please rewrite the above conditions using multiple early-exit `if`s.

Majkl578 commented on lib/DoctrineCodingStandard/Sniffs/Classes/ExceptionInterfaceNamingSniff.php:
```diff
@@ -0,0 +1,97 @@
+        }
+
+        $phpcsFile->addError(
+            'Exception interface needs an "Exception" name suffix',
```
_must end with_

Majkl578 commented on lib/DoctrineCodingStandard/Sniffs/Classes/ExceptionInterfaceNamingSniff.php:
```diff
@@ -0,0 +1,97 @@
+
+        if ($hasExceptionName && ! $isExtendingException && ! $isExtendingThrowable) {
+            $phpcsFile->addError(
+                'Interface does not extend an exception interface',
```
_must extend_

Majkl578 commented on lib/DoctrineCodingStandard/Sniffs/Classes/ExceptionClassNamingSniff.php:
```diff
@@ -0,0 +1,99 @@
+        return $isImplementingExceptions || $isImplementingThrowable;
+    }
+
+    private function hasExceptionSuffix(string $className) : bool
```
Please extract into Helpers, reused on multiple places / sniffs.

Majkl578 commented on lib/DoctrineCodingStandard/Sniffs/Classes/ExceptionClassNamingSniff.php:
```diff
@@ -0,0 +1,99 @@
+    /**
+     * {@inheritdoc}
+     */
+    public function process(File $phpcsFile, $stackPtr) : void
```
`$pointer`

Majkl578 commented on lib/DoctrineCodingStandard/Sniffs/Classes/ExceptionClassNamingSniff.php:
```diff
@@ -0,0 +1,99 @@
+    public function process(File $phpcsFile, $stackPtr) : void
+    {
+        $isAbstract              = $phpcsFile->findFirstOnLine([T_ABSTRACT], $stackPtr) !== false;
+        $isFinal                 = $phpcsFile->findFirstOnLine([T_FINAL], $stackPtr) !== false;
```
Since class can't be both abstract and final, you this should preferably use `SlevomatCodingStandard\Helpers\TokenHelper::findPreviousEffective($phpcsFile, $pointer)` and check for `T_FINAL`/`T_ABSTRACT` (using this instead of `findFirstOnLine()` is also safer).

Majkl578 commented on lib/DoctrineCodingStandard/Sniffs/Classes/ExceptionClassNamingSniff.php:
```diff
@@ -0,0 +1,99 @@
+            return false;
+        }
+
+        $isImplementingExceptions = count(preg_grep('/Exception$/', $implementedInterfaces)) > 0;
```
Please extract into Helpers, reused on multiple places / sniffs.

Majkl578 commented on lib/DoctrineCodingStandard/Sniffs/Classes/ExceptionClassNamingSniff.php:
```diff
@@ -0,0 +1,99 @@
+        }
+
+        $isImplementingExceptions = count(preg_grep('/Exception$/', $implementedInterfaces)) > 0;
+
```
You should early-exit here if `$isImplementingExceptions` is true.

SenseException replied:
This is the sniff for classes. Implementing an Exception interface is not enough for a class, because it needs to extend another Exception class too. I don't know about a requirement that an Exception class has to implement an interface.

Majkl578 replied:
>  Implementing an Exception interface is not enough for a class

That's true, but this sniff is about naming only, it should not report inheritance issues IMO.
Not extending some Exception will cause PHP fatal error, but that is out of scope for this sniff.

SenseException replied:
I even start to see no point here of checking for an exception interface when a class, regular or abstract, has to extend an exception. This is a relict from before making exception classes `final`. I think the complexity of checks for exception interfaces can be removed here. How do you think about that?

Majkl578 commented on lib/DoctrineCodingStandard/Sniffs/Classes/ExceptionClassNamingSniff.php:
```diff
@@ -0,0 +1,99 @@
+
+        $importedClassNames = UseStatementHelper::getUseStatements($phpcsFile);
+
+        // TODO Should throwable be checked separately, because it can't be implemented on non-abstract exception class?
```
Let's not allow this at all - you can't implement Throwable without extending Exception/Error.

Majkl578 commented on lib/DoctrineCodingStandard/Sniffs/Classes/ExceptionClassNamingSniff.php:
```diff
@@ -0,0 +1,99 @@
+        $hasValidClassName       = ($isAbstract && $hasExceptionName) ||
+            (! $isAbstract && ! $hasExceptionName && $isFinal);
+        $isValidException        = $hasValidClassName && ($isExtendingException || $isImplementingException);
+        $isNoException           = ! $hasExceptionName && ! $isExtendingException && ! $isImplementingException;
```
Please try to turn these combinations into multiple early-exit `if`s, this is too complex.

Majkl578 commented on lib/DoctrineCodingStandard/Sniffs/Classes/ExceptionClassNamingSniff.php:
```diff
@@ -0,0 +1,99 @@
+
+        if (! $hasValidClassName) {
+            $phpcsFile->addError(
+                'Use "Exception" suffix for abstract exception classes and make non-abstract classes final',
```
_Exception classes must end with an "Exception" suffix._ (should not mention abstract/final here since we check for class name)

Majkl578 commented on lib/DoctrineCodingStandard/Sniffs/Classes/ExceptionClassNamingSniff.php:
```diff
@@ -0,0 +1,99 @@
+        }
+
+        $phpcsFile->addError(
+            'Class is not a valid exception',
```
Should say why: _Exception classes must be either abstract or final._

Majkl578 reviewed (changes requested):
I did a more thorough review - let me know if something is not clear, reviewing sniffs isn't easy since different people choose different approaches for the same outcome. :)

SenseException committed:
Renaming of variables and exception texts

SenseException committed:
Move logic of private methods into helper classes

SenseException committed:
Use early exit for class name checks

SenseException committed:
Update return type in docs for register method

SenseException committed:
Fix of check for valid class name

SenseException commented:
Not ready yet, but I'm almost done.

SenseException committed:
Use Slevomat helper to get abstract or final token

SenseException committed:
Handle classes that aren't extending exceptions

SenseException committed:
Rearrange of the variable rows in the class naming sniff

SenseException committed:
Replace explode on token-string with token parser logic

SenseException commented:
Ready for another review.

alcaeus commented on .travis.yml:
```diff
@@ -20,6 +20,8 @@ install: travis_retry composer update --prefer-dist
 script:
   - vendor/bin/phpcs
   - vendor/bin/phpcs $(find tests/input/* | sort) --report=summary --report-file=phpcs.log; diff tests/expected_report.txt phpcs.log
+  - vendor/bin/phpstan analyse
```
These should be put into separate stages like we do on other projects.

SenseException replied:
Other Doctrine projects? I'll take a look into that.

alcaeus commented on phpcs.xml.dist:
```diff
@@ -13,7 +13,8 @@
     <arg value="p"/>
 
-    <rule ref="Doctrine"/>
+    <rule ref="DoctrineCodingStandard"/>
```
That is quite the BC break - are we sure we have to/want to rename the standard?

SenseException replied:
I run into issues and chose the second solution mentioned:
https://github.com/doctrine/coding-standard/pull/63#issuecomment-404022002

But I admit, I didn't use much time to test how to make it run with the previous namespace.

alcaeus commented on tests/test/ExceptionClassNamingSniffNotValid.php:
```diff
@@ -0,0 +1,35 @@
+
+final class StillNoExtension implements FooException
+{
+}
```
Some of these files don't contain a trailing EOL. It's either missing from the sniff or we don't properly check tests. Either way, there should be a consistent rule on this.

SenseException replied:
These tests are only for the new exception sniffs to prevent failing tests on new additions in future PRs.

alcaeus replied:
I mean the trailing newlines - this is part of PSR-2 and should be fixed here as well.

alcaeus commented on lib/DoctrineCodingStandard/Helpers/UseStatementHelper.php:
```diff
@@ -0,0 +1,37 @@
+    public static function isImplementingThrowable(array $importedClassNames, array $implementedInterfaces) : bool
+    {
+        return (in_array(Throwable::class, $importedClassNames, true) &&
+            in_array(Throwable::class, $implementedInterfaces, true)) ||
```
Just to make sure, I assume this could lead to issues if `Throwable` is aliased?

```
use My\Namespace\Throwable;
use Throwable as BaseThrowable;

class Exception implements Throwable
```

We might want to also fetch aliases from `use` statements and explicitly check against that to avoid wrong results.

SenseException replied:
Like this one?
https://github.com/doctrine/coding-standard/pull/63/files#diff-fea81a7c92d7e638eff90b9ea3554855R36

alcaeus replied:
I guess the whole thing depends on whether `findImplementedInterfaceNames` returns aliased names or the actual FQCN. If it's the latter, this works as expected.

FWIW, there's a typo in `TrowableAlias` in your example that I just spotted now.

SenseException replied:
Can you please point to the typo?

Majkl578 committed:
Fix standard name & namespace

Majkl578 committed:
Improve exception naming sniffs

Majkl578 committed:
Exclude redundant SlevomatCodingStandard.Classes.SuperfluousExceptionNaming

SenseException committed:
Fix of typo of interface name

SenseException committed:
Fix of merged and rebased code changes

SenseException committed:
Use hardcoded expected sniffer error code in test

The code also is used to disable sniffs in projects.
Therefore changing a code constant would be a BC break and should be
catched by the test.

SenseException commented:
This PR should also get an addition to the documentation for the new rules.

SenseException commented:
phpcbf seems to change the docblock of tests/input2/test-case.php from

```
    /**
     * @test
     *
     * @covers MyClass::test
     *
     * @uses MyClass::__construct
     */
```
to
```
    /**
     * @uses MyClass::__construct
     *
     * @test
     * @covers MyClass::test
     */
```

The Apply fixes stage build fails.

SenseException commented:
I've rebased to the current branch to keep this PR up to date.

ostrolucky commented:
I agree with @Ocramius, not liking idea about any exception classes being final.

On different note, even though this PR seems like great work, it does too much rules at once, which is why people hesitate to review it. Having less rules at least initially would allow less disagreements and easier time to review and merge it.

Ocramius commented:
@SenseException do you believe this to be ready for review or still WIP?

SenseException commented:
I would prefer reviews for the changes @Majkl578 and I did. Feedback can help development on this PR. It does a lot of stuff and there might be still some uncovered cases. The question about having `final` a few months ago is still open, which started in Slack and wasn't discussed here much. This is a contribution that involves first custom code for this repository after all.

SenseException commented:
I would prefer reviews before I rebase against master.

SenseException commented on lib/Doctrine/Helpers/NamingHelper.php:
```diff
@@ -0,0 +1,41 @@
+{
+    public static function hasExceptionSuffix(string $className) : bool
+    {
+        return preg_match('~Exception$~', $className) === 1;
```
Interfaces have to end with `Exception`. I can add `ExceptionInterface` to a test for a fail case.

ostrolucky replied:
> Interfaces have to end with Exception

That's true only for doctrine code. Userland (or doctrine codebase itself) can be extending 3rd party exceptions with different naming convention. Detecting Exception(Interface)? is small step but easy win in such cases

SenseException replied:
True. It started for Doctrine code, but if someone else wants to use it, this might become a nuisance. But the detection of what is an Exception and what not depends on a naming convention. Your example uses a classic "end with Interface" convention, but what about a 3rd party exception naming convention without any suffixes or prefixes? This would make detecting exceptions with this sniff very hard.

What about making the naming convention configurable?

ostrolucky replied:
If you want to make this configurable, feel free. For me, handling ExceptionInterface would be good enough step, as such convention is common.

SenseException commented on lib/Doctrine/Helpers/UseStatementHelper.php:
```diff
@@ -0,0 +1,61 @@
+    public static function getRealName(File $phpcsFile, string $referencedName) : string
+    {
+        $uses = self::getUseStatements($phpcsFile);
+        return $uses[$referencedName] ?? $referencedName;
```
That rule had to be added after my last commit, because the last build is green. After review, I'll rebase and fix the rest of the coding styles.

Ocramius commented:
Deferred to `8.0.0`

carusogabriel commented:
What is the status of this PR?

SenseException commented:
I'm currently not working on this WIP and don't know when I'll come back to it.

SenseException commented:
I'm currently not working on this WIP and don't know when I'll come back to it.

SenseException commented:
I'm currently not working on this WIP and don't know when I'll come back to it.
